### PR TITLE
Fix for JAMES-2263

### PR DESCRIPTION
**Pull request by apptaro**

### CODE REVIEW

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/camel/CamelProcessor.java [hunk 1]
```diff
@@ -70,7 +68,7 @@ public void process(Exchange exchange) throws Exception {
         Mail mail = exchange.getIn().getBody(Mail.class);
         long start = System.currentTimeMillis();
         TimeMetric timeMetric = metricFactory.timer(mailet.getClass().getSimpleName());
-        MessagingException ex = null;
```
Comment (chibenwa) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/camel/CamelProcessor.java]: How is the `ex` variable usefull?

Can't we get a ride of it?

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/camel/MatcherSplitter.java [hunk 2]
```diff
@@ -120,7 +120,9 @@
                     ProcessorUtil.verifyMailAddresses(matchedRcpts);
                 }
 
-            } catch (MessagingException me) {
+            } catch (IOException e) {
```
Comment (chibenwa) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/camel/MatcherSplitter.java]: What is specific about IOException? Why can't it be handled as a normal exception?

File: server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java [hunk 3]
```diff
@@ -201,7 +202,7 @@ public void afterMailet(Mailet m, String mailName, String state, long processTim
         } catch (MessagingException e) {
             catched = true;
         }
-        assertTrue(catched);
+        assertFalse(catched);
```
Comment (chibenwa) [inline server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java]:
```
boolean catched = false;
try {
    // do something
} catch(Exception e) {
    catched = true;
}
assertFalse(catched);
```

Is strictly equivalent to : 

```
// do something
```

(and the second options prints error details, easing debugging...)

File: server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java [hunk 4]
```diff
@@ -255,7 +256,7 @@ public void afterMailet(Mailet m, String mailName, String state, long processTim
...
```
Comment (chibenwa) [inline server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java]: Idem...

File: server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java [hunk 5]
```diff
@@ -211,7 +212,7 @@ public void afterMailet(Mailet m, String mailName, String state, long processTim
     }
 
     @Test
-    public void testMailetThrowException() throws Exception {
+    public void testMailetNoThrowException() throws Exception {
```
Comment (chibenwa) [inline server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java]: That test name look possible to improve and should match our naming convention ;-)

File: server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java [hunk 6]
```diff
@@ -163,7 +164,7 @@ public void afterMailet(Mailet m, String mailName, String state, long processTim
     }
 
     @Test
-    public void testMatcherThrowException() throws Exception {
+    public void testMatcherNoThrowException() throws Exception {
```
Comment (chibenwa) [inline server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java]: We generally tend to name the tests we (re) write like this:

```
methodShouldAction[WhenCondition]
```

Here very likely:

```
matcherProcessingShouldNotResultInAnExceptionWhenMatcherThrows
```

**Comment by chibenwa** [comment]
You still need to improve you commit messages.

 - 1. It needs to be prefixed by the ticket number
 - 2. You need a message to explain what you did

Here I would propose:

```
JAMES-2263 Correct mailetContainer error handling
```

**Comment by chibenwa** [comment]
You might also want to cherry-pick this commit: https://github.com/chibenwa/james-project/commit/484c423c4a60febd6b442d4b7210656244220012

It proposes integration tests for mailet pipeline errors. You can remove the @Ignore in it to check your fix ;-)

**Comment by chibenwa** [comment]
You might also want to cherry-pick this commit: https://github.com/chibenwa/james-project/commit/484c423c4a60febd6b442d4b7210656244220012

It proposes integration tests for mailet pipeline errors. You can remove the @Ignore in it to check your fix ;-)

**Comment by apptaro** [comment]
Thank you very much for your quick response and everything. Sorry for the commit message. I read about it after I made the commit. I'm looking forward to the new release with the fixes.

**Comment by chibenwa** [comment]
No problem.

Pro tip: Use `git rebase` `reword` feature for correcting the commit message.

**Comment by apptaro** [comment]
Do you know when this fix is included in the release?

**Comment by chibenwa** [comment]
@apptaro  When merged the fixes are available on the master branch. You could be using it while waiting for a fix.

Concerning a release date, that is a good question. We need to switch to time based release so that I'm confident answering it.

Cheers,

**Comment by chibenwa** [comment]
Hi,

Any news about this pull request? Would you find time for handling the above comment?

Cheers, merry Christmas and happy new year!

Benoit

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/camel/CamelProcessor.java (hunk 1 above)
Comment (apptaro) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/camel/CamelProcessor.java]: The `ex` variable is used in the `finally` clause, so we can't get rid of it.

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/camel/MatcherSplitter.java (hunk 2 above)
Comment (apptaro) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/camel/MatcherSplitter.java]: I'm not sure what is specific about IOException. The original code handles it differently, so I left it as it is, but I think onMatchException setting should be honored in the same way when IOException occurs, so I will remove it.

File: server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java (hunk 6 above)
Comment (apptaro) [reply server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java]: Okay, I will change that, and do the same for testMailetNoThrowException.

File: server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java (hunk 3 above)
Comment (apptaro) [reply server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java]: Okay, I will change that.

File: server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java (hunk 5 above)
Comment (apptaro) [reply server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java]: I will change it.

File: server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java (hunk 4 above)
Comment (apptaro) [reply server/mailet/mailetcontainer-camel/src/test/java/org/apache/james/mailetcontainer/lib/AbstractStateMailetProcessorTest.java]: I will change it.

**Comment by apptaro** [comment]
Oops, I have done all the fixing, but I have added unnecessary files to a commit. What's the easiest way to fix this? Sorry I'm fairly new to git..

**Comment by chibenwa** [comment]
No problem, it is quite easy:

```
git log path/to/file
```

Will give you the SHA-1 identifier of the commit containing the mistake.

```
git rebase -i master
```

Will then allow you to rework the history of your branch.

Edit that specific commit. You can do that by prefixing it by a `e` in the rebase view.

```
e c785e20a7a JAMES-2263 My awesome commit
```

Then you can reset the commit:

```
git reset --soft HEAD~
```

You can then remove the file you don't want to include anymore:

```
git reset file/I/do/not/want/to/commit.txt
```

When done, re-commit the changes:

```
git commit -m "JAMES-2263 My awesome commit"
```

And finish the rebase with

```
git rebase --continue
```

If git complains about an empty commit, perform that command to skip the empty commit:

```
git rebase --skip
```

That's maybe a lot of commands, but it ends up not that hard to  do, when you get the habit.

**Comment by chibenwa** [comment]
Pro tip: Position a `.gitignore` to prevent you from committing these files in the first place.

Mine is:

```
target
.idea
*.iml
.gitignore
keystore
*.lib
*.jar
/site/
/.m2/
/release/
KahaDB/
src/homepage/_site/
src/homepage/.sass-cache/
src/homepage/Gemfile.lock
dependency-reduced-pom.xml
*.bak
.floo*
```

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/camel/CamelProcessor.java (hunk 1 above)
Comment (chibenwa) [reply server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/camel/CamelProcessor.java]: Ok, got it. Let's live with it...

**Comment by apptaro** [comment]
Hi, thank you very much. I think this PR is clean now. Please take a look.

**Comment by chibenwa** [comment]
Hi,

I also created an other ticket for : 
 - Attaching the encountered exception to the mail
 - Have a HasException matcher (that will allow matching the FQDN of the exception)

This will allow very finned grained error handling while not having to write custom mailet/matcher.

As this is related to this contribution, I thought you might be interested. So I am mentioning the link here:

https://issues.apache.org/jira/browse/JAMES-2275

This proposal of contribution is on a related topic, should not be too hard to implement, and might even interest you ;-)

Cheers,

Benoit Tellier

**Comment by chibenwa** [comment]
Hi,

This contribution was just merged. Could you close this pull request?

Many thanks again for that fix!

Cheers, and see you around.

Benoit Tellier

**Comment by apptaro** [comment]
@chibenwa Thank you very much!

**Pull request closed without merging**